File: backend/app/engine/phases.py
"""阶段枚举与转移逻辑：夜间阶段序、角色到阶段映射、expected_actors。"""

from __future__ import annotations

from enum import StrEnum


class Phase(StrEnum):
    LOBBY = "LOBBY"
    ROLE_ASSIGN = "ROLE_ASSIGN"
    NIGHT_GUARD = "NIGHT_GUARD"
    NIGHT_WEREWOLF = "NIGHT_WEREWOLF"
    NIGHT_WITCH = "NIGHT_WITCH"
    NIGHT_SEER = "NIGHT_SEER"
    NIGHT_HUNTER_CONFIRM = "NIGHT_HUNTER_CONFIRM"
    WIN_CHECK = "WIN_CHECK"
    SHERIFF_ELECTION = "SHERIFF_ELECTION"
    SHERIFF_PK = "SHERIFF_PK"
    DEATH_ANNOUNCE = "DEATH_ANNOUNCE"
    LAST_WORDS = "LAST_WORDS"
    DAY_SPEECH = "DAY_SPEECH"
    VOTE = "VOTE"
    VOTE_PK = "VOTE_PK"
    EXILE = "EXILE"
    HUNTER_SHOOT = "HUNTER_SHOOT"
    IDIOT_FLIP = "IDIOT_FLIP"
    GAME_OVER = "GAME_OVER"


from app.engine.config import GameConfig, RoleType  # noqa: E402
from app.engine.state import (  # noqa: E402
    GameState,
    living,
    living_of_role,
    living_wolves,
)

_ROLE_TO_NIGHT_PHASE: dict[RoleType, Phase] = {
    RoleType.GUARD: Phase.NIGHT_GUARD,
    RoleType.WEREWOLF: Phase.NIGHT_WEREWOLF,
    RoleType.WITCH: Phase.NIGHT_WITCH,
    RoleType.SEER: Phase.NIGHT_SEER,
    RoleType.HUNTER: Phase.NIGHT_HUNTER_CONFIRM,
}


def phase_for_role(role: RoleType) -> Phase | None:
    """返回角色对应的夜间阶段；无夜间行动的角色返回 None。"""
    return _ROLE_TO_NIGHT_PHASE.get(role)


def night_phase_sequence(config: GameConfig) -> list[Phase]:
    """按 config.night_order 过滤出有对应夜间阶段的角色序列（去重）。"""
    seq: list[Phase] = []
    for role in config.night_order:
        ph = phase_for_role(role)
        if ph is not None and ph not in seq:
            seq.append(ph)
    return seq


def next_night_phase(config: GameConfig, current: Phase) -> Phase | None:
    """序列中 current 的下一夜间阶段；末尾或未找到返回 None。"""
    seq = night_phase_sequence(config)
    if current not in seq:
        return seq[0] if seq else None
    idx = seq.index(current)
    return seq[idx + 1] if idx + 1 < len(seq) else None


def expected_actors(state: GameState) -> set[int]:
    """当前必须行动的座位集合；系统阶段返回空集。"""
    ph = state.phase

    if ph == Phase.NIGHT_GUARD:
        return {
            p.seat for p in living_of_role(state, RoleType.GUARD) if p.seat not in state.acted_seats
        }
    if ph == Phase.NIGHT_WEREWOLF:
        return {w.seat for w in living_wolves(state) if w.seat not in state.wolf_proposals}
    if ph == Phase.NIGHT_WITCH:
        return {
            w.seat
            for w in living_of_role(state, RoleType.WITCH)
            if w.seat not in state.acted_seats and (w.witch_antidote or w.witch_poison)
        }
    if ph == Phase.NIGHT_SEER:
        return {
            p.seat for p in living_of_role(state, RoleType.SEER) if p.seat not in state.acted_seats
        }
    if ph == Phase.NIGHT_HUNTER_CONFIRM:
        # M1：猎人首夜确认无决策，作为系统直通阶段（can_shoot 在死亡时判定）
        return set()

    if ph == Phase.DAY_SPEECH:
        if state.speech_idx < len(state.speech_order):
            return {state.speech_order[state.speech_idx]}
        return set()

    if ph == Phase.VOTE:
        return {p.seat for p in living(state) if p.can_vote and p.seat not in state.votes}
    if ph == Phase.VOTE_PK:
        return {
            p.seat
            for p in living(state)
            if p.can_vote and p.seat not in state.vote_candidates and p.seat not in state.votes
        }

    if ph == Phase.HUNTER_SHOOT:
        return {state.pending_hunter} if state.pending_hunter is not None else set()

    if ph == Phase.LAST_WORDS:
        if state.speech_idx < len(state.speech_order):
            return {state.speech_order[state.speech_idx]}
        return set()

    if ph == Phase.SHERIFF_ELECTION:
        if state.election_stage == "candidacy":
            return {p.seat for p in living(state) if p.seat not in state.sheriff_declared}
        if state.election_stage == "vote":
            return {
                p.seat
                for p in living(state)
                if p.can_vote
                and p.seat not in state.sheriff_candidates
                and p.seat not in state.sheriff_votes
            }
        if state.election_stage == "direction":
            return {state.sheriff_seat} if state.sheriff_seat is not None else set()
        return set()
    if ph == Phase.SHERIFF_PK:
        return {
            p.seat
            for p in living(state)
            if p.can_vote
            and p.seat not in state.sheriff_candidates
            and p.seat not in state.sheriff_votes
        }

    # 其余为系统阶段
    return set()

File: frontend/src/engine/phases.ts
// 中文文案表：对齐 backend/app/cli/render.py（_ROLE_ZH、_ELECTION_STAGE_ZH、render_event）
// 与 backend/app/engine/events.py（BadgeLostReason 注释）。纯常量 + 纯函数，零 IO。

import type { ElectionStage, Phase, RoleType } from "./types";

export const PHASE_ZH: Record<Phase, string> = {
  LOBBY: "大厅",
  ROLE_ASSIGN: "发牌",
  NIGHT_GUARD: "夜晚 · 守卫行动",
  NIGHT_WEREWOLF: "夜晚 · 狼人行动",
  NIGHT_WITCH: "夜晚 · 女巫行动",
  NIGHT_SEER: "夜晚 · 预言家行动",
  NIGHT_HUNTER_CONFIRM: "夜晚 · 猎人确认",
  WIN_CHECK: "胜负判定",
  SHERIFF_ELECTION: "白天 · 警长竞选",
  SHERIFF_PK: "白天 · 警长 PK",
  DEATH_ANNOUNCE: "白天 · 公布死讯",
  LAST_WORDS: "白天 · 遗言",
  DAY_SPEECH: "白天 · 发言",
  VOTE: "白天 · 投票",
  VOTE_PK: "白天 · PK 投票",
  EXILE: "白天 · 放逐",
  HUNTER_SHOOT: "猎人开枪",
  IDIOT_FLIP: "白痴翻牌",
  GAME_OVER: "游戏结束",
};

export const ELECTION_STAGE_ZH: Record<ElectionStage, string> = {
  "": "竞选环节结束",
  candidacy: "上警报名",
  speech: "上警发言",
  withdraw: "退水确认",
  vote: "警下投票",
  direction: "警长决定发言方向",
  announce: "公布结果",
};

export const ROLE_ZH: Record<RoleType, string> = {
  WEREWOLF: "狼人",
  VILLAGER: "村民",
  SEER: "预言家",
  WITCH: "女巫",
  HUNTER: "猎人",
  GUARD: "守卫",
  IDIOT: "白痴",
};

export const ROLE_ABBR: Record<RoleType, string> = {
  WEREWOLF: "狼",
  SEER: "预",
  WITCH: "巫",
  HUNTER: "猎",
  GUARD: "守",
  IDIOT: "痴",
  VILLAGER: "民",
};

// 与 events.py::BadgeLostReason 的中文注释一致。
export const BADGE_LOST_ZH: Record<string, string> = {
  NO_CANDIDATES: "无人上警",
  ALL_WITHDREW: "候选人全员退水",
  NO_VOTERS: "全员上警，无警下投票人",
  TIE_AGAIN: "PK 再平票",
  SELF_DESTRUCT: "竞选期狼人自爆吞警徽",
};

// 与 render.py::render_event 的 GAME_OVER 分支一致（winner 为 null 时另行显示「平局」）。
export const WINNER_ZH: Record<string, string> = {
  GOOD: "好人阵营",
  WOLF: "狼人阵营",
};

const NIGHT_PHASES: ReadonlySet<Phase> = new Set([
  "NIGHT_GUARD",
  "NIGHT_WEREWOLF",
  "NIGHT_WITCH",
  "NIGHT_SEER",
  "NIGHT_HUNTER_CONFIRM",
]);

export function isNight(phase: Phase): boolean {
  return NIGHT_PHASES.has(phase);
}

/** 角色所属阵营配色变量名（Nocturne tokens）：狼/神职/民。 */
export function factionColorVar(role: RoleType): string {
  if (role === "WEREWOLF") return "--ah-wolf";
  if (role === "VILLAGER" || role === "IDIOT") return "--ah-villager";
  return "--ah-god"; // SEER / WITCH / HUNTER / GUARD
}
